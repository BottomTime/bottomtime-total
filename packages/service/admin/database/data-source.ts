import path from 'path';
import { DataSource } from 'typeorm';

export async function getDataSource(
  url: string,
  requireSsl: boolean,
): Promise<DataSource> {
  const client = new DataSource({
    type: 'postgres',
    url,
    entities: [path.resolve(__dirname, '../../src/data/**/*.entity.ts')],
    ssl: requireSsl,
  });
  return await client.initialize();
}
