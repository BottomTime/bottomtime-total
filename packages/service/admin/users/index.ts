import { CommandModule } from 'yargs';

import { createAdmin } from './create-admin';
import { getUserToken } from './get-user-token';

export const userModule: CommandModule<{
  'postgres-uri': string;
  'require-ssl': boolean;
}> = {
  command: 'user',

  describe: 'Commands for managing a user account',

  builder(yargs) {
    return yargs
      .command(
        'new-admin',
        'Creates a new admin user.',
        (yargs) => {
          return yargs
            .positional('username', {
              alias: 'u',
              description: "The user's username.",
              type: 'string',
            })
            .positional('password', {
              alias: 'p',
              description: 'The password to set on the new admin account',
              type: 'string',
            })
            .help();
        },
        async (yargs) => {
          await createAdmin({
            postgresRequireSsl: yargs.requireSsl,
            potgresUri: yargs.postgresUri,
            username: yargs.username,
            password: yargs.password,
          });
        },
      )
      .command(
        'token <user>',
        'Gets a token for a user',
        (yargs) => {
          return yargs
            .positional('user', {
              demandOption: true,
              description: "The user's username or email address.",
              type: 'string',
            })
            .help();
        },
        async (yargs) => {
          await getUserToken(yargs.postgresUri, yargs.requireSsl, yargs.user);
        },
      )
      .help();
  },

  async handler() {
    /* TODO */
  },
};
