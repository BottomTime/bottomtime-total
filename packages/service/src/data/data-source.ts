import path from 'path';
import { DataSource, DataSourceOptions } from 'typeorm';

import { Config } from '../config';

export const PostgresDataSourceOptions: DataSourceOptions = {
  type: 'postgres',
  url: Config.postgresUri,
  entities: [path.join(__dirname, './**/*.entity.ts')],
  synchronize: false,
};

export async function initDataSource(): Promise<DataSource> {
  const AppDataSource = new DataSource(PostgresDataSourceOptions);
  return await AppDataSource.initialize();
}
