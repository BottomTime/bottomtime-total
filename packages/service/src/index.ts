import { Config } from './config';
import { createApp } from './create-app';
import { createLogger } from './logger';

const log = createLogger(Config.logLevel);

createApp(log)
  .then(async (app) => {
    await app.listen(Config.port);

    log.info(
      `ðŸŽ‰ Service has successfully started and is listening on port ${Config.port}. ðŸŽ‰`,
    );
  })
  .catch((error) => {
    log.fatal('[SERVICE] Failed to start service... Shutting down...', error);
    process.exit(1);
  });
