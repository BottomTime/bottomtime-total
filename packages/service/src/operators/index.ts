export * from './operator';
export * from './operator-review';

export * from './operators.module';
