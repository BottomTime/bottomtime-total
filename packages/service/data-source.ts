import path from 'path';
import { DataSource } from 'typeorm';

import { PostgresDataSourceOptions } from './src/data/data-source';

export default new DataSource({
  ...PostgresDataSourceOptions,
  migrations: [path.join(__dirname, './migrations/*.ts')],
});
