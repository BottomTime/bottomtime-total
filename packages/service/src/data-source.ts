import 'reflect-metadata';
import { DataSourceOptions } from 'typeorm';

import { Config } from './config';
import * as Entities from './data';

export const PostgresDataSourceOptions: DataSourceOptions = {
  type: 'postgres',
  url: Config.postgresUri,
  entities: Object.values(Entities),
  synchronize: false,
};
