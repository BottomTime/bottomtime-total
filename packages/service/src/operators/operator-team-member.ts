import { TeamMemberDTO } from '@bottomtime/api';

import { Repository } from 'typeorm';

import { OperatorTeamMemberEntity } from '../data';
import { User, UserFactory } from '../users';

export class OperatorTeamMember {
  constructor(
    private readonly members: Repository<OperatorTeamMemberEntity>,
    private readonly userFactory: UserFactory,
    private readonly data: OperatorTeamMemberEntity,
  ) {}

  get title(): string | undefined {
    return this.data.title ?? undefined;
  }
  set title(value: string | undefined) {
    this.data.title = value ?? null;
  }

  get joined(): string | undefined {
    return this.data.joined ?? undefined;
  }
  set joined(value: string | undefined) {
    this.data.joined = value ?? null;
  }

  get member(): User {
    if (!this.data.teamMember) {
      throw new Error('Team member not loaded');
    }
    return this.userFactory.createUser(this.data.teamMember);
  }

  async save(): Promise<void> {
    await this.members.save(this.data);
  }

  async delete(): Promise<void> {
    await this.members.delete(this.data.id);
  }

  toJSON(): TeamMemberDTO {
    return {
      title: this.title,
      joined: this.joined?.valueOf(),
      member: this.member.profile.toJSON(),
    };
  }
}
