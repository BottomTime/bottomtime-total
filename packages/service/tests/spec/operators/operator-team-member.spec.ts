import { AccountTier, LogBookSharing } from '@bottomtime/api';

import { Repository } from 'typeorm';

import {
  OperatorEntity,
  OperatorTeamMemberEntity,
  UserEntity,
} from '../../../src/data';
import { OperatorTeamMember } from '../../../src/operators/operator-team-member';
import { UserFactory } from '../../../src/users';
import { dataSource } from '../../data-source';
import { createTestOperator, createTestUser } from '../../utils';

describe('OperatorTeamMember class', () => {
  let Users: Repository<UserEntity>;
  let Operators: Repository<OperatorEntity>;
  let Members: Repository<OperatorTeamMemberEntity>;
  let userFactory: UserFactory;

  let user: UserEntity;
  let operator: OperatorEntity;
  let data: OperatorTeamMemberEntity;
  let teamMember: OperatorTeamMember;

  beforeAll(() => {
    Users = dataSource.getRepository(UserEntity);
    Operators = dataSource.getRepository(OperatorEntity);
    Members = dataSource.getRepository(OperatorTeamMemberEntity);

    userFactory = new UserFactory(Users);

    user = createTestUser({
      id: '9d1cf2fc-e112-43dd-9584-5800477c2af2',
      bio: 'OMG! Diving!!',
      accountTier: AccountTier.Basic,
      avatar: 'https://example.com/avatar.jpg',
      name: 'John Doe',
      memberSince: new Date('2021-01-01T00:00:00Z'),
      experienceLevel: 'Expert',
      logBookSharing: LogBookSharing.FriendsOnly,
      location: 'San Francisco, CA',
      username: 'Diver McDiverface',
      startedDiving: '2008',
    });
    operator = createTestOperator(user);
  });

  beforeEach(async () => {
    data = {
      id: '4e237902-8ab2-40b2-b744-8b78b2261cf3',
      operator,
      teamMember: user,
      title: 'Lead Instructor',
      joined: '2021-01-01',
    };

    await Users.save(user);
    await Operators.save(operator);

    teamMember = new OperatorTeamMember(Members, userFactory, data);
  });

  it('will return properties correctly', () => {
    expect(teamMember.title).toBe(data.title);
    expect(teamMember.joined).toEqual(data.joined);
  });

  it('will return null properties correctly', () => {
    data.title = null;
    data.joined = null;
    expect(teamMember.title).toBeUndefined();
    expect(teamMember.joined).toBeUndefined();
  });

  it('will render as JSON', () => {
    expect(teamMember.toJSON()).toMatchSnapshot();
  });

  it('will save a new team member', async () => {
    await teamMember.save();
    const saved = await Members.findOneOrFail({
      where: { id: data.id },
      relations: ['teamMember', 'operator'],
    });
    expect(saved.title).toBe(data.title);
    expect(saved.joined).toEqual(data.joined);
    expect(saved.teamMember?.id).toBe(user.id);
    expect(saved.operator?.id).toBe(operator.id);
  });

  it('will update an existing team member', async () => {
    await Members.save(data);

    const newTitle = 'Dive Master';
    const newJoined = '2024-04-04';
    teamMember.title = newTitle;
    teamMember.joined = newJoined;
    await teamMember.save();

    const saved = await Members.findOneOrFail({
      where: { id: data.id },
      relations: ['teamMember', 'operator'],
    });
    expect(saved.title).toBe(newTitle);
    expect(saved.joined).toEqual(newJoined);
    expect(saved.teamMember?.id).toBe(user.id);
    expect(saved.operator?.id).toBe(operator.id);
  });

  it('will delete a team member', async () => {
    await Members.save(data);
    await teamMember.delete();
    await expect(Members.findOneBy({ id: data.id })).resolves.toBeNull();
  });

  it('will do nothing if deleting a team member that does not exist', async () => {
    await teamMember.delete();
  });
});
