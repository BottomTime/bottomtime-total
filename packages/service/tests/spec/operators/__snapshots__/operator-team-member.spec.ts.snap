// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OperatorTeamMember class will render as JSON 1`] = `
{
  "joined": "2021-01-01",
  "member": {
    "accountTier": 0,
    "avatar": "https://example.com/avatar.jpg",
    "bio": "OMG! Diving!!",
    "experienceLevel": "Expert",
    "location": "San Francisco, CA",
    "logBookSharing": "friends",
    "memberSince": 1609459200000,
    "name": "John Doe",
    "startedDiving": "2008",
    "userId": "9d1cf2fc-e112-43dd-9584-5800477c2af2",
    "username": "Diver McDiverface",
  },
  "title": "Lead Instructor",
}
`;
